LOG4J2-3628 Replace `maven-changes-plugin` with `log4j-changelog`

vy opened the pull request:
### What is what?
* `src/changelog` – where changelog _sources_ are stored
* `src/changelog/.<releaseVersionMajor>.x.x` – changelog sources for the upcoming release
* `src/changelog/<releaseVersion>` – changelog sources for the past released versions
* `src/changelog/<releaseVersion>/.release.xml` – the meta information about the release (e.g., release version and date)
* `src/changelog/<releaseVersion>/.changelog.adoc.ftl` – the FreeMarker-templated AsciiDoc-formatted release notes used by the generator (manually populated from past announcement emails)
* `src/changelog/*/[<issueId>_]<shortSummary>.xml` – changelog entry associated with a change (manually populated from existing `changes.xml`)
* `target/generated-sources/site/asciidoc/changelog/<releaseVersion>.adoc` – generated via `log4j-changelog`
* `target/generated-sources/site/asciidoc/changelog/index.adoc` – generated via `log4j-changelog`
* `log4j-changelog` – changelog importer (from `changes.xml`), AsciiDoc exporter, changelog releaser

### How shall a reviewer start?
* Proof-read `CHANGELOG.adoc`
* Quickly skim through `changelog/*/*.{xml,adoc}` files, since they are either generated or copied

### What is next?
1. Clone this to `master`
1. Clean duplicate `changes.xml` entries there (:heavy_check_mark:)

ppkarwasz commented on log4j-internal-util/src/main/java/org/apache/logging/log4j/internal/util/changelog/exporter/AsciiDocExporter.java:
```diff
@@ -0,0 +1,424 @@
+
+    private static List<ChangelogEntry> readChangelogEntries(final Path releaseDirectory) {
+        return findAdjacentFiles(releaseDirectory)
+                .sorted(FILE_MODIFICATION_TIME_COMPARATOR)
```
Are we sure that `git clone` sets the correct modification time?

vy replied:
That is a really good point @ppkarwasz! I don't know and I think it indeed not a good idea to rely on that. I have fixed this in d64770b1ca056a69b472759a1048a340838d83fc by ordering changelog entries by filepath.

ppkarwasz reviewed (changes requested):
LGTM, although I am curious why didn't you use Jackson or JAXB for serialization and deserialization of XML.

vy commented on log4j-internal-util/pom.xml:
```diff
@@ -0,0 +1,164 @@
+              <goal>verify</goal>
+            </goals>
+          </execution>
+        </executions>
```
Fixed in 583050c1c917c786a78e461a49b4a2be1e22d66c.

vy commented on log4j-internal-util/src/main/java/org/apache/logging/log4j/internal/util/XmlReader.java:
```diff
@@ -0,0 +1,120 @@
+        final SAXParser parser = parserFactory.newSAXParser();
+        final DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
+        final DocumentBuilder documentBuilder = documentBuilderFactory.newDocumentBuilder();
+        final Document document = documentBuilder.newDocument();
```
Fixed in 583050c1c917c786a78e461a49b4a2be1e22d66c.

vy commented:
> LGTM, although I am curious why didn't you use Jackson or JAXB for serialization and deserialization of XML.

@ppkarwasz, there are a couple of reasons:

* AFAIK, both Jackson and JAXB doesn't decently support comment blocks needed for license headers.
* Jackson and XML attributes have always been troublesome for me.
* I wanted `log4j-internal-util` to be dependency-free so that if needed one can simple do a `java -cp target/log4j-internal-util.jar ...` – though I need to admit this is a problem that can easily be addressed via [`spring-boot-maven-plugin`'s `repackage` goal](https://docs.spring.io/spring-boot/docs/current/maven-plugin/reference/htmlsingle/#goals-repackage).
* At the early stages of development, I started with JAXB and later on made a Jackson attempt too. Yet after failing a couple of times, I decided to bite the bullet.

For me the current setup simply works. Though above arguments might be due to my lack of knowledge. I am not proud of shipping `XmlReader`, `XmlWriter`, `XmlUtils`, etc. classes either. :sweat_smile: If you think this can be done simpler via Jackson/JAXB, feel free to either give it a shot or tell me how to do it.

garydgregory commented:
Note that the XML spec does not require parsers to provide comments as part of a parsed model:

https://www.w3.org/TR/xml/#sec-comments

vy commented:
> Note that the XML spec does not require parsers to provide comments as part of a parsed model:
> 
> https://www.w3.org/TR/xml/#sec-comments

@garydgregory, thanks, good to know. Though it is irrelevant for this PR, since we don't need to parse comments, but emit them while generating XML.

rgoers commented:
There are a couple of things I am not crazy about with this

1. Why is the profile to run this in the internal-utils project. It makes more sense to be in the parent pom and automatically included as part the the "apache-release" profile which is enabled automatically by the release plugin. 
2. Why isn't the source for the changelog in src/changelog?

rgoers reviewed (commented):
Note that I can't really review this PR properly as my browser never completes loading all the files.

In addition to my previous comments I don't really see a good way to test this as the tool fails unless a release has been set up - i.e. I can't tell it to generate a changes.xml for the current SNAPSHOT.

vy commented:
Thanks so much for your feedback @rgoers! Before addressing your remarks, let me note the following changes I have incorporated in the meantime:

1. Added `src/site/.gitignore` to ignore `src/site/changelog`. Yes, AsciiDoc-formatted changelogs won't be committed anymore. Since this content is only relevant for the _generated_ website and it is a major source of merge-conflicts.
1. Renamed `.unreleased` directory to `.unreleased-<releaseVersionMajor>.x.x`. This allows multiple unreleased version changelogs (e.g., `.unreleased-2.x.x`, `.unreleased-3.x.x`) to coexist, which is necessary for my next PRs to fuse the website for both `release-2.x` and `master` branches.

> Why is the profile to run this in the internal-utils project. It makes more sense to be in the parent pom and automatically included as part the the "apache-release" profile which is enabled automatically by the release plugin.

I have a couple of reasons for the current approach where `log4j-internal-util` module is used to host changelog business logic:

* I doubt if it is an idiomatic Maven practice to have Java sources in a `pom`-typed artifact.
* [`pom`-typed artifacts don't support Java sources out of the box.](https://maven.apache.org/ref/3.8.6/maven-core/artifact-handlers.html)
* Exporter preferably should run during `package` phase (rather than in a certain release profile) so that if somebody breaks changelog files (e.g., due to malformed XML) this should break the build too.

> 2. Why isn't the source for the changelog in src/changelog?

Done.

> Note that I can't really review this PR properly as my browser never completes loading all the files.

I would suggest checking out the branch to your local development environment and reviewing the changes there. Grouping the changes by directory helps a lot.

> In addition to my previous comments I don't really see a good way to test this as the tool fails unless a release has been set up - i.e. I can't tell it to generate a changes.xml for the current SNAPSHOT.

I see. I have changed this behaviour. `./mvnw -pl log4j-internal-util -P changelog-releaser` should work with `*-SNAPSHOT` versions now.

rgoers commented:
@vy I didn't suggest that you move the source from log4j-internal-util to the parent project. I suggested that the plugin invocation be in the parent pom. However, that may mean that the utility not be part of the log4j build as it will require that project be built first. That actually makes sense to me since the utility doesn't really need to be publicly released.

I will re-pull the branch and test it again.

rgoers commented:
I don't understand the latest changes. When running it I see
```
moving `/Users/rgoers/projects/apache/logging/log4j/release-2.x/src/changelog/.unreleased-2.x.x` to `/Users/rgoers/projects/apache/logging/log4j/release-2.x/src/changelog/20221128-2.19.1-SNAPSHOT`
writing release information to `/Users/rgoers/projects/apache/logging/log4j/release-2.x/src/changelog/20221128-2.19.1-SNAPSHOT/.release.xml`
created intro file at `/Users/rgoers/projects/apache/logging/log4j/release-2.x/src/changelog/20221128-2.19.1-SNAPSHOT/.intro.adoc
```

1. Why is it moving stuff from the unreleased directory. Won't that cause it to be lost on a subsequent invocation?
2. Why does the target have a date if everything is moved? Why wouldn't it just be 2.19.1-SNAPSHOT? 
3. Isn't there is a risk that these SNAPSHOT directories will actually be committed? 
4. I didn't see an index.adoc generated anywhere.

To be honest, I am still not comfortable with this. 
1. Why is it necessary to move files?  I would think renaming the .unreleased directory to the release number would be sufficient. This should happen as part of release:prepare.
2. Generated files are not normally added to the source but are generated under the target directory. If the tool performs step 1 and then generates a changes.adoc file with changes for all releases similar to what the maven changes plugin does then I would hope the changes.adoc file could be generated to the target directory and included by asciidoc from there.

ppkarwasz commented:
@vy,

The remark about SAX what just a curiosity. I usually use StaX as "modern" equivalent of SAX with both XML parsing and writing. However IIRC Xalan does not support DOM-to-StaX transformations and the OpenJDK fork of Xalan depends on the JDK, so SAX is probably the safest choice.

vy commented:
Thanks so much for your remarks @rgoers! :pray: Let me respond to them.

> I didn't suggest that you move the source from log4j-internal-util to the parent project. I suggested that the plugin invocation be in the parent pom. However, that may mean that the utility not be part of the log4j build as it will require that project be built first. That actually makes sense to me since the utility doesn't really need to be publicly released. In fact, in thinking about this I think you should move log4j-internal-util to the logging-log4j-tools repo as log4j-changes. That is really what that repo is for. I'm not sure what we should do with log4j-server though. It really isn't a tool. Maybe we need a logging-log4j-samples repo.

I see your point. If these changelog utilities would be used by other projects besides Log4j, it would even make more sense to move it to a dedicated repository. Though, as of now, given we haven't even used it for once yet, don't you agree that it is a bit overkill to move these utilities to a separate (rusted?) repository, make a release there, so that we can use it in Log4j... _just because_ they should be invoked from the parent project rather than a module?

I can agree with this plan, but please, let's first use this a couple of times in Log4j, polish it, etc. Then it would be trivial to move this to a separate repository.

> I will re-pull the branch and test it again.

:bow:

> Why is it moving stuff from the unreleased directory. Won't that cause it to be lost on a subsequent invocation?

You are making a release (e.g., `2.20.0`), not a release candidate. The entire `.unreleased-2.x.x` will be converted to `20??????-2.20.0`, just like your `2.20.0-SNAPSHOT` artifacts have been converted to `2.2.20` and there is no turning back.

> Why does the target have a date if everything is moved? Why wouldn't it just be 2.19.1-SNAPSHOT?

It provides an ordering in `src/changelog` that is more convenient for developers, IMHO. Nevertheless, date is not necessary for folder names, though it is necessary for populating the `.release.xml` file.

> Isn't there is a risk that these SNAPSHOT directories will actually be committed?

That was the reason I did not allow `*-SNAPSHOT` versions in `ChangelogReleaser` earlier, but since you wanted to experiment... Anyway, recently added 86610b36ce10adc8874baa233c7b4fd9e0d8c8d0 disallows SNAPSHOT versions unless `log4j.changelog.releaser.snapshotsAllowed` Java property is defined, e.g., `./mvnw -Dlog4j.changelog.releaser.snapshotsAllowed -pl log4j-internal-util -P changelog-releaser`.

> I didn't see an index.adoc generated anywhere.

It works for me:

```
$ ./mvnw -Dlog4j.changelog.releaser.snapshotsAllowed -pl log4j-internal-util -P changelog-releaser
...
using `20221128` and `2.19.1-SNAPSHOT` for release date and version, respectively
moving `/home/vy/Projects/log4j/2/src/changelog/.unreleased-2.x.x` to `/home/vy/Projects/log4j/2/src/changelog/20221128-2.19.1-SNAPSHOT`
writing release information to `/home/vy/Projects/log4j/2/src/changelog/20221128-2.19.1-SNAPSHOT/.release.xml`
created intro file at `/home/vy/Projects/log4j/2/src/changelog/20221128-2.19.1-SNAPSHOT/.intro.adoc`

$ cat /home/vy/Projects/log4j/2/src/changelog/20221128-2.19.1-SNAPSHOT/.intro.adoc
////
    Licensed to the Apache Software Foundation (ASF) under one or more
    contributor license agreements.  See the NOTICE file distributed with
    this work for additional information regarding copyright ownership.
    The ASF licenses this file to You under the Apache License, Version 2.0
    (the "License"); you may not use this file except in compliance with
    the License.  You may obtain a copy of the License at

         https://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
////
```

> Why is it necessary to move files?  I would think renaming the .unreleased directory to the release number would be sufficient.

That is exactly what we are doing: `mv .unreleased-2.x.x 20221128-2.19.1`, even the `ChangelogReleaser` tells so:

```
moving `/home/vy/Projects/log4j/2/src/changelog/.unreleased-2.x.x` to `/home/vy/Projects/log4j/2/src/changelog/20221128-2.19.1`
```

> This should happen as part of release:prepare.

The reason I did not want to bind this to any existing _release_ phase is to avoid generating (and hence sealing up!) the changelog for release candidates. I expect the release manager to consciously seal the changelog just like (s)he does for the ultimate release.

> Generated files are not normally added to the source but are generated under the target directory. If the tool performs step 1 and then generates a changes.adoc file with changes for all releases similar to what the maven changes plugin does then I would hope the changes.adoc file could be generated to the target directory and included by asciidoc from there.

That is right, ideally generated files should be placed under `target/generated-sources/<i-dunno-what-exactly>`. Though [`asciidoctor-maven-plugin`'s `maven-site-plugin` integration doesn't allow adding extra source paths](https://docs.asciidoctor.org/maven-tools/latest/site-integration/setup-and-configuration/#setup).

rgoers commented:
> I see your point. If these changelog utilities would be used by other projects besides Log4j, it would even make more sense to move it to a dedicated repository. Though, as of now, given we haven't even used it for once yet, don't you agree that it is a bit overkill to move these utilities to a separate (rusted?) repository, make a release there, so that we can use it in Log4j... just because they should be invoked from the parent project rather than a module?
No, I don't agree. If we intend to use it from a different repo we need to do that. As I said, the plugin should be referenced in the parent pom. We can't do that if the code is in a sub-module.
> You are making a release (e.g., 2.20.0), not a release candidate. The entire .unreleased-2.x.x will be converted to 20??????-2.20.0, just like your 2.20.0-SNAPSHOT artifacts have been converted to 2.2.20 and there is no turning back.
All releases are release candidates. They only become a release when the vote passes and the release tag is created.  It is not uncommon to add a new Jira to a subsequent release candidate and we cannot be precluded from doing it.
>| I didn't see an index.adoc generated anywhere.
>It works for me:
You are showing it generating .intro.adoc, not index.adoc as the README says. I see no file generated that contains all the changelog entries, which is what I thought the main point of this was.
